modules/service/src/main/scala/ch/epfl/bluebrain/nexus/iam/service/auth/DownstreamAuthClient.scala

```diff
@@ -51,18 +51,31 @@ class DownstreamAuthClient[F[_]](config: OidcConfig, cl: UntypedHttpClient[F], u
     forward(Get(config.tokenEndpoint.withQuery(Query("code" -> code, "state" -> state))))
 
   /**
-    * Fetches and forwards the [[UserInfo]] object associated to these ''credentials''.
+    * Forwards the ''userinfo'' object received from the provider for these ''credentials''.
     * @param credentials the OAuth 2.0 bearer token given by the provider
-    * @return a [[UserInfo]] instance in an ''F'' context
+    * @return the response in an ''F'' context
+    */
+  def userInfo(credentials: OAuth2BearerToken): F[HttpResponse] =
+    forward(Get(config.userinfoEndpoint).addCredentials(credentials))
+
+  /**
+    * Fetches the ''userinfo'' associated to this access token and builds the corresponding [[User]] instance.
+    * @param accessToken the OAuth 2.0 bearer token given by the provider
+    * @return a future [[User]] holding all the identities the user belongs to
     */
-  def userInfo(credentials: OAuth2BearerToken): F[UserInfo] = {
-    uicl(Get(config.userinfoEndpoint).addCredentials(credentials))
+  def getUser(accessToken: String): F[User] = {
+    uicl(Get(config.userinfoEndpoint).addCredentials(OAuth2BearerToken(accessToken)))
+      .map(_.toUser(config.issuer))
       .recoverWith {
         case df: DecodingFailure =>
           log.error("Unable to decode UserInfo response", df)
-          F.raiseError(df)
+          F.raiseError(UnexpectedUnsuccessfulHttpResponse(HttpResponse(StatusCodes.BadGateway)))
         case UnexpectedUnsuccessfulHttpResponse(resp) =>
-          log.error("Received a failed response from the downstream oid provider for getting the UserInfo")
+          if (resp.status == StatusCodes.Unauthorized) {
```

Comment:
Can we do pattern matching here instead of `if`, e.g. something along the lines of 
```
case UnexpectedUnsuccessfulHttpResponse(HttpResponse(StatusCodes.Unauthorized, _, _, ....) => (...)
case UnexpectedUnsuccessfulHttpResp(resp) => (....)
(..)
```

Reply:
I thought pattern matching would be unnecessarily clunky, that's why I used an `if`